### PR TITLE
RUMM-1322 Allow initializing `DDURLSessionDelegate` before starting the SDK

### PR DESCRIPTION
**Pull request by ncreated**
### What and why?

🚚 This PR removes the constraint of initializing `DDURLSessionDelegate` after `Datadog.initialize()`. From now on, users can create our delegate freely, no matter of `Datadog.initialize()`. The SDK will only track requests issued after starting the SDK.

This is to provide more freedom for the user setup and cut off error-prone edge cases (like race conditions).

### How?

I changed the `DDURLSessionDelegate` to obtain the interceptor's reference through lazily computed variable, instead of injecting its value on `init()`. The obsolete misconfiguration warnings were also removed.

To test the new behaviour, I randomized the `URLSession`-based integration tests setup to randomly create delegate before or after the `Datadog.initialize()`. This includes 4 integration tests using `URLSession` and `NSURLSession`: 2 for RUM Resources and 2 for Tracing auto instrumentation.

### Review checklist

- [x] Feature or bugfix MUST have appropriate tests (unit, integration)
- [x] Make sure each commit and the PR mention the Issue number or JIRA reference

### CODE REVIEW

File: Sources/Datadog/URLSessionAutoInstrumentation/Interception/URLSessionInterceptor.swift [hunk 1]
```diff
@@ -6,18 +6,29 @@
...
     func taskCompleted(task: URLSessionTask, error: Error?)
+
+    var handler: URLSessionInterceptionHandler { get }
 }
 
 /// An object performing interception of requests sent with `URLSession`.
 public class URLSessionInterceptor: URLSessionInterceptorType {
     public static var shared: URLSessionInterceptor? {
-        URLSessionAutoInstrumentation.instance?.interceptor
+        URLSessionAutoInstrumentation.instance?.interceptor as? URLSessionInterceptor
```
Comment (ncreated) [inline Sources/Datadog/URLSessionAutoInstrumentation/Interception/URLSessionInterceptor.swift:31]: 💡 This casting is now required and is [tested in existing `URLSessionAutoInstrumentationTests`](https://github.com/DataDog/dd-sdk-ios/blob/4044b7d65f0a86ef3149ba7f5be58d49275d3567/Tests/DatadogTests/Datadog/URLSessionAutoInstrumentation/URLSessionAutoInstrumentationTests.swift#L21) tests.

File: Sources/Datadog/URLSessionAutoInstrumentation/Interception/URLSessionInterceptor.swift [hunk 2]
```diff
@@ -6,18 +6,29 @@
...
+
+/// An interface for processing `URLSession` task interceptions.
 internal protocol URLSessionInterceptorType: class {
     func modify(request: URLRequest, session: URLSession?) -> URLRequest
     func taskCreated(task: URLSessionTask, session: URLSession?)
     func taskMetricsCollected(task: URLSessionTask, metrics: URLSessionTaskMetrics)
     func taskReceivedData(task: URLSessionTask, data: Data)
     func taskCompleted(task: URLSessionTask, error: Error?)
+
+    var handler: URLSessionInterceptionHandler { get }
```
Comment (ncreated) [inline Sources/Datadog/URLSessionAutoInstrumentation/Interception/URLSessionInterceptor.swift:25]: 💡 This adds a bit noise to the `URLSessionInterceptorType` interface and could be additionally mitigated by decoupling the `handler` creation to upstream (to `URLSessionAutoInstrumentation` class). This however seems over engineering if done in the scope of this PR - we may want to refactor this on a stronger reason with more constraints for changing the abstraction.

File: Datadog/Example/Scenarios/URLSession/URLSessionScenarios.swift [hunk 3]
```diff
@@ -14,9 +14,13 @@ import Datadog
...
-    /// If yes, instrumented endpoints are passed to `DDURLSessionDelegate`; otherwise, they are passed to `DatadogConfiguration.trackURLSession` method
-    @objc
-    let feedAdditionalFirstyPartyHosts: Bool
+    /// Randomizes the way of passing additional first party hosts.
+    /// If `true`, instrumented endpoints are passed to `DDURLSessionDelegate`; otherwise, they are passed to `DatadogConfiguration.trackURLSession(...)`.
+    private let feedAdditionalFirstyPartyHosts: Bool
+
+    /// Randomizes the way of creating `URLSession` instrumented with `DDURLSessionDelegate`.
+    /// If `true`, the session is created after `Datadog.initialize()`; if `false`, it's created before.
+    private let lazyInitRLSession: Bool
```
Comment (buranmert) [inline Datadog/Example/Scenarios/URLSession/URLSessionScenarios.swift:23]: typo
```suggestion
    private let lazyInitURLSession: Bool
```

File: Sources/Datadog/URLSessionAutoInstrumentation/DDURLSessionDelegate.swift [hunk 4]
```diff
@@ -28,44 +28,22 @@ open class DDURLSessionDelegate: NSObject, URLSessionTaskDelegate, URLSessionDat
         return self
     }
 
-    var interceptor: URLSessionInterceptorType?
-    let firstPartyURLsFilter: FirstPartyURLsFilter?
+    internal var interceptor: URLSessionInterceptorType? { URLSessionAutoInstrumentation.instance?.interceptor }
+    private(set) var firstPartyURLsFilter: FirstPartyURLsFilter?
```
Comment (buranmert) [inline Sources/Datadog/URLSessionAutoInstrumentation/DDURLSessionDelegate.swift:32]: why not let?

**Review by buranmert — APPROVED** [review]

**Pull request merged**